README
This is basically the plugin by weasel5i2, all credits to him. I just added support for Minecraft 1.2.3 and switched to PermissionsEx

Wings - Simple Flight
Version: v1.4

This plugin lets you fly around. Simply hold a feather in your hand and right-click with it to toggle flying and hovering on/off. First click will enable flight mode. A second right-click will switch you into hover mode. A third right-click will switch off flying and you will fall to the ground (without damage, as long as you're still holding the feather!)

You will fly in the direction you're facing, so if you're going to hit that mountain, well.. PULL UP!! PULL UP!!!

Features:
Flying around in the sky like some crazy bird.
Hovering! (mostly..) - while hovering, hold down SHIFT to move yourself upwards.
Permissions node:
- 'wings.fly'

Configuration:
The configuration is located in plugins/Wings/Settings.ini. The file looks like this:

hoverMultiplier=0.0012410253
hoverBoostMultiplier=0.00555
allowOps=true
flyingEatsFeathers=true
defaultFeatherAmount=500
flightSpeed=1.0

* hoverMultiplier: The vector amount to achieve the player hovering effect. Your mileage may vary depending on how this is set.
* hoverBoostMultiplier: The vector amount to get that extra upward boost when you press SHIFT whilst hovering.
* allowOps: Allow Op users to fly, regardless of their Permissions setting.
* flightSpeed: The amount to multiply the flight vector by. Don't set this too high or you may end up extremely far away from your intended destination!
* flyingEatsFeathers: (disabled by default) - Set whether or not flying and hovering will consume feathers. Note that the feather must be held in the FIRST hand inventory slot #1 (at least until I can figure out how to get it to see them elsewhere in the player's inventory)
* defaultFeatherAmount: This sets how many ticks each feather will last before being consumed. On my server, the default 500 setting makes them last approximately 15-20 seconds each before disappearing from the stack.


CHANGELOG
Changelog:

Version 1.4 10-Mar-2012
* Added support for Minecraft 1.2.3
* Added support for PermissionsEx (instead of Permissions)
* Changed max flightheight to 300

Version 1.3 04-Apr-2011
* Added feather consuming support
* Added configuration file plugins/Wings/Settings.ini (see above)

Version 1.2 03-Apr-2011
* Added Permissions 2.5.5 support
* Added hovering!

Version 1.1 31-Mar-2011 - unreleased
* Experimental stuff. Didn't work out too well. 

Version 1.0 23-Mar-2011
* Created plugin


src/net/weasel/Wings/TimerTask.java
package net.weasel.Wings;

import net.weasel.Wings.Wings.flyingState;

import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.Server;
import org.bukkit.World;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.bukkit.util.Vector;

public class TimerTask implements Runnable
{
	Wings plugin = null;
	Server server = null;
	
	public TimerTask( Wings instance )
	{
		plugin = instance;
		server = plugin.getServer();
	}
	
	@Override
	public void run() 
	{
//		double M = Wings.hoverMultiplier;
		
		World world = null;
		Player player = null;

		for( int W = 0; W < server.getWorlds().size(); W++ )
		{
			world = server.getWorlds().get(W);
			
			for( int P = 0; P < world.getPlayers().size(); P++ )
			{
				player = world.getPlayers().get(P);
				
				if( Wings.isFlying( player ) == flyingState.HOVERING )
				{
/*					if( player.isSneaking() == true ) M = Wings.hoverBoostMultiplier;
					
					double X = Wings.getHoverLocation(player).getX();
					double Y = Wings.getHoverLocation(player).getY();
					double Z = Wings.getHoverLocation(player).getZ();
					
					Vector pVector = new Vector( X, Y , Z );
					Vector hVector = new Vector( 0, M, 0 );
					
					Vector hover = pVector.multiply( hVector );
					hover.clone();
					
					player.setVelocity( new Vector( 0,0,0 ) );
					player.setVelocity( hover );*/

					player.setVelocity( new Vector( 0,0,0 ) );
					
					Location last = Wings.getHoverLocation(player);
					Location current = player.getLocation();
					last.setPitch(current.getPitch());
					last.setYaw(current.getPitch());
					
					player.teleport(last);
					
					
					if( Wings.flyingEatsFeathers == true && Wings.isFlying(player) != flyingState.NOT_FLYING )
					{
						Integer fAmount = Wings.getFeatherPoints(player);
						
						fAmount--;
						
						Wings.setFeatherPoints( player, fAmount );
						
						if( fAmount < 1 )
						{
							try
							{
								if( player.getInventory().all(Material.FEATHER).get(0).getAmount() < 1 )
								{
									System.out.println( "out of feathers (OPM)");
									Wings.setFlying( player, flyingState.NOT_FLYING );
									player.sendMessage( "You are no longer flying." );
								}
								else
								{
									ItemStack F = new ItemStack( Material.FEATHER, 1 );
									player.getInventory().removeItem(F);
	
									if( player.getInventory().all(Material.FEATHER).get(0).getAmount() < 1 )
									{
										Wings.setFlying( player, flyingState.NOT_FLYING );
										player.sendMessage( "You are no longer flying." );
									}
									else
									{
										Wings.setFeatherPoints( player, Wings.defaultFeatherAmount );
									}
								}

								Wings.setFeatherPoints( player, Wings.defaultFeatherAmount );
							}
							catch( Exception e )
							{
								Wings.setFlying( player, flyingState.NOT_FLYING );
								player.sendMessage( "You are no longer flying." );
							}
						}
					}
				}
			}
		}
	}
}

src/net/weasel/Wings/EntityListener.java
package net.weasel.Wings;

//import javax.persistence.EntityListeners;

import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.event.entity.EntityDamageEvent;
import org.bukkit.event.Listener;
import org.bukkit.event.EventHandler;

public class EntityListener implements Listener//extends org.bukkit.event.entity.EntityListener 
{
	public static Wings plugin;
	
	public EntityListener( Wings instance ) 
	{
		plugin = instance;
	}

	// Prevent damage when allowed to fly
	@EventHandler
	public void onEntityDamage(EntityDamageEvent event)
	{
	    if ((event.getCause().equals(EntityDamageEvent.DamageCause.FALL)) && ((event.getEntity() instanceof Player)))
	    {
		    try 
		    {
		    	Player player = ((Player)event.getEntity());
		    	
		    	if( Wings.isAllowed(player) )
		    	{
		    		if( Wings.falldamageWithoutFeather && player.getItemInHand().getType() != Material.FEATHER )
		    			return;
		
		    		event.setCancelled( true );
		    	}
		    }
		    catch (Exception e) 
		    {
		    	// Do nothing..
		    }
	    }
	}
}


src/net/weasel/Wings/PlayerListener.java
package net.weasel.Wings;

import net.weasel.Wings.Wings.flyingState;

import org.bukkit.Material;
import org.bukkit.Server;
import org.bukkit.entity.Player;
import org.bukkit.event.block.Action;
import org.bukkit.event.player.PlayerInteractEvent;
import org.bukkit.event.player.PlayerMoveEvent;
import org.bukkit.event.player.PlayerQuitEvent;
import org.bukkit.inventory.ItemStack;
import org.bukkit.util.Vector;

import org.bukkit.event.Listener;
import org.bukkit.event.EventHandler;

public class PlayerListener implements Listener // extends org.bukkit.event.player.PlayerListener 
{
	public Wings plugin;
	public Server server;
	
	public PlayerListener(Wings instance)
	{
		plugin = instance;
		server = plugin.getServer();
	}

	// Allow player flying around depending on state
	@EventHandler
	public void onPlayerMove( PlayerMoveEvent event )
	{
		Player player = event.getPlayer();
		
		if( Wings.isFlying( player ) == flyingState.FLYING )
		{
			if( player.getLocation().getY() < 300 )
			{
				Vector dir = player.getLocation().getDirection();
				player.setVelocity( dir.multiply(Wings.flightSpeed) );
			}
		}

		if( Wings.isFlying( player ) == flyingState.HOVERING )
		{
			double M = Wings.hoverMultiplier;
			
			if( player.isSneaking() == true ) M = Wings.hoverBoostMultiplier;
			
			double X = Wings.getHoverLocation(player).getX();
			double Y = Wings.getHoverLocation(player).getY();
			double Z = Wings.getHoverLocation(player).getZ();
			
			Vector pVector = new Vector( X, Y , Z );
			Vector hVector = new Vector( 0, M, 0 );
			
			Vector hover = pVector.multiply( hVector );
			hover.clone();
			
			player.setVelocity( new Vector( 0,0,0 ) );
//			player.setVelocity( hover );
			
			player.teleport(Wings.getHoverLocation(player));
		}
		
		if( Wings.flyingEatsFeathers == true && Wings.isFlying(player) != flyingState.NOT_FLYING )
		{
			Integer fAmount = Wings.getFeatherPoints(player);
			
			fAmount--;
			
			Wings.setFeatherPoints( player, fAmount );
			
			if( fAmount < 1 )
			{
				try
				{
					if( player.getInventory().all(Material.FEATHER).get(0).getAmount() < 1 )
					{
						System.out.println( "out of feathers (OPM)");
						Wings.setFlying( player, flyingState.NOT_FLYING );
						player.sendMessage( "You are no longer flying." );
					}
					else
					{
						ItemStack F = new ItemStack( Material.FEATHER, 1 );
						player.getInventory().removeItem(F);
		
						if( player.getInventory().all(Material.FEATHER).get(0).getAmount() < 1 )
						{
							Wings.setFlying( player, flyingState.NOT_FLYING );
							player.sendMessage( "You are no longer flying." );
						}
						else
						{
							Wings.setFeatherPoints( player, Wings.defaultFeatherAmount );
						}
					}
		
					Wings.setFeatherPoints( player, Wings.defaultFeatherAmount );
				}
				catch( Exception e )
				{
					Wings.setFlying( player, flyingState.NOT_FLYING );
					player.sendMessage( "You are no longer flying." );
				}
			}
		}
	}
	
	// player using feather
	@EventHandler
	public void onPlayerInteract( PlayerInteractEvent event )
	{
		Player player = event.getPlayer();
		
		if( Wings.isAllowed(player) == true )
		{
			if( player.getItemInHand().getType() == Material.FEATHER )
			{
				if( event.getAction() == Action.RIGHT_CLICK_AIR || 
					event.getAction() == Action.RIGHT_CLICK_BLOCK )
				{
					if( Wings.isFlying(player) == flyingState.NOT_FLYING )
					{
						player.sendMessage( "You are now flying." );
						
						if( Wings.featherPoints.containsKey(player) == false )
							Wings.setFeatherPoints( player, Wings.defaultFeatherAmount );
						
						Wings.setFlying( player, flyingState.FLYING );
					}
					else if( Wings.isFlying(player) == flyingState.FLYING )
					{
						Wings.setHoverLocation( player );
						player.setVelocity( new Vector( 0,0,0 ) );
						player.sendMessage( "You are now hovering." );
						Wings.setFlying( player, flyingState.HOVERING );
					}
					else
					{
						player.sendMessage( "You are no longer flying." );
						Wings.setFlying( player, flyingState.NOT_FLYING );
					}
	
					event.setCancelled( true );
				}
			}
		}
	}

	// clean up when player disconnects
	@EventHandler
	public void onPlayerQuit( PlayerQuitEvent event )
	{
		if( Wings.flyingPlayers.containsKey(event.getPlayer() ) )
		{
			Wings.flyingPlayers.remove( event.getPlayer() );
			System.out.println( "[Wings] Player quit; removed from list." );
		}
	}
}


src/net/weasel/Wings/Wings.java
package net.weasel.Wings;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.FileWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintStream;
import java.io.PrintWriter;
import java.util.HashMap;

import org.bukkit.Location;
import org.bukkit.Server;
//import org.bukkit.plugin.Plugin;
//import org.bukkit.plugin.PluginManager;
import org.bukkit.plugin.java.JavaPlugin;
import org.bukkit.scheduler.BukkitScheduler;
import org.bukkit.command.Command;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;

import ru.tehkode.permissions.PermissionManager;
import ru.tehkode.permissions.bukkit.PermissionsEx;
//import org.bukkit.entity.Player;
//import org.bukkit.event.EventPriority;
//import org.bukkit.event.*;
/*import com.nijiko.permissions.PermissionHandler;
import com.nijikokun.bukkit.Permissions.Permissions;*/


public class Wings extends JavaPlugin 
{
	public static enum flyingState {
		NOT_FLYING,
		FLYING,
		HOVERING,
	}
	
	public final static HashMap<Player,flyingState> flyingPlayers = new HashMap<Player,flyingState>();
	public final static HashMap<Player,Location> hoverLocs = new HashMap<Player,Location>();
	public final static HashMap<Player,Integer> featherPoints = new HashMap<Player,Integer>();
	
	public static int tTask = 0;
	public static BukkitScheduler timer = null;
	public static Wings plugin;
	public static Server server;
	public static PermissionManager Permissions;
	public static double flightSpeed;
	public static double hoverMultiplier;
	public static double hoverBoostMultiplier;
	public static boolean allowOps;
	public static boolean flyingEatsFeathers;
	public static Integer defaultFeatherAmount;
	public static boolean falldamageWithoutFeather;
	public static String iniFile = "plugins/Wings/Settings.ini";
	
	@Override
	public void onDisable() 
	{
		try
		{
			timer.cancelTask( tTask );
		}
		catch( Exception e )
		{
			e.getMessage();
		}
		
		System.out.println( "[Wings] Wings v" + this.getDescription().getVersion() + " disabled." );
	}

	// initialize
	@Override
	public void onEnable() 
	{
		checkConfig();
		setupPermissions();
		
		plugin = this;
//		server = getServer();
//		Priority priority = Event.Priority.Monitor;
		hoverMultiplier = getDblSetting( "hoverMultiplier", 0.0012410253 );
		hoverBoostMultiplier = getDblSetting( "hoverBoostMultiplier", 0.00555 );
		allowOps = getBooleanSetting( "allowOps", true );
		flyingEatsFeathers = getBooleanSetting( "flyingEatsFeathers", false );
		defaultFeatherAmount = getIntSetting( "defaultFeatherAmount", 100 );
		flightSpeed = getDblSetting( "flightSpeed", 1.0 );
		falldamageWithoutFeather = getBooleanSetting( "falldamageWithoutFeather", false );
		timer = getServer().getScheduler();
		
//		PluginManager pm = server.getPluginManager();
		PlayerListener pListener = new PlayerListener(this);
		EntityListener eListener = new EntityListener(this);
		TimerTask task = new TimerTask(plugin);
		
		getServer().getPluginManager().registerEvents(pListener, plugin);
		getServer().getPluginManager().registerEvents(eListener, plugin);
/*		pm.registerEvent(Event.Type.PLAYER_QUIT, pListener, EventPriority.MONITOR, plugin );
		pm.registerEvent(Event.Type.PLAYER_MOVE, pListener, EventPriority.MONITOR, plugin );
		pm.registerEvent(Event.Type.PLAYER_INTERACT, pListener, EventPriority.MONITOR, plugin );
		pm.registerEvent(Event.Type.ENTITY_DAMAGE, eListener, EventPriority.MONITOR, plugin );*/
	
		tTask = timer.scheduleSyncRepeatingTask( plugin, task, 1, 1 );
		
		System.out.println( "[Wings] Wings v" + this.getDescription().getVersion() + " loaded." );
		System.out.println( "[Wings] Hover multiplier set to " + hoverMultiplier );
		System.out.println( "[Wings] Hover boost multiplier set to " + hoverBoostMultiplier );
		System.out.println( "[Wings] Flying consumes feathers: " + ( flyingEatsFeathers ? "TRUE" : "FALSE" ) );
		
		if( flyingEatsFeathers == true )
			System.out.println( "[Wings] Default per-feather amount to consume: " + defaultFeatherAmount );
			
		System.out.println( "[Wings] Ops allowed to fly regardless: " + ( allowOps ? "TRUE" : "FALSE" ) );
	}
	
    public boolean onCommand(CommandSender sender, Command command, String label, String[] args )
	{
		return false;
	}
    
    public static void checkConfig()
    {
    	File iniDir = new File( "plugins/Wings" );
    	
    	if( iniDir.exists() == false )
    	{
	    	if( iniDir.mkdir() == true )
	    		System.out.println( "[Wings] Created Settings.ini folder: plugins/Wings/" );
    	}
    	
    	File ini = new File(iniFile);
    	
    	if( ini.exists() == false )
    	{
    		if( createIniFile() == true )
    		{
    			System.out.println( "New Settings.ini file created with default settings." );
    		}
    	}
    	
    }
	
	public static flyingState isFlying(final Player player)
	{
		if(flyingPlayers.containsKey(player))
			return flyingPlayers.get(player);
		else
			return flyingState.NOT_FLYING;
	}

	public static Location getHoverLocation(final Player player)
	{
		if(hoverLocs.containsKey(player))
			return hoverLocs.get(player);
		else
			return player.getLocation();
	}

	public static void setFlying(final Player player, final flyingState value)
	{
		flyingPlayers.put(player,value);
	}

	public static void setHoverLocation(final Player player )
	{
		hoverLocs.put(player,player.getLocation());
	}

	public static void setFeatherPoints(final Player player, int points )
	{
		featherPoints.put(player,points);
	}

	public static Integer getFeatherPoints(final Player player )
	{
		if( featherPoints.containsKey(player) == true )
			return( featherPoints.get(player) );
		else
		{
			featherPoints.put( player, defaultFeatherAmount );
			return( defaultFeatherAmount );
		}
	}

	public static void logOutput( String text )
	{
		System.out.println( "[Wings] " + text );
	}

	private void setupPermissions() 
	{
		if (Wings.Permissions == null) 
	    {
		    if(getServer().getPluginManager().isPluginEnabled("PermissionsEx"))
			{
	    		Wings.Permissions = PermissionsEx.getPermissionManager();
		    }
		    else 
	    	{
	    		logOutput("Permission system not detected, defaulting to OP");
	        }
	    }
	}

	public static boolean isAllowed( Player player )
	{
		if( Permissions == null )
		{
			return true;
		}
		
		if( Permissions.has(player, "wings.fly"))
		{
		    return true;
		}
		else
		{
			if( player.isOp() == true && allowOps == true )
			{
				return true;
			}
			else
			{
				return false;
			}
		}
	}

    public String[] getSetting( String which, String Default )
    {
        return( getSettingValue(iniFile, which, Default, "" ) );
    }

    public boolean getBooleanSetting( String which, boolean dValue )
    {
    	boolean retVal;
    	
    	String dString = ( dValue ? "true" : "false" );
    	
        if( getSettingValue(iniFile, which, dString, "" )[0].equalsIgnoreCase("true") )
        	retVal = true;
        else
        	retVal = false;
        
        return( retVal );
    }
    
	public String[] getSettingValue(String fileName, String optionName, String defaultValue, String splitValue)
    {
    	Boolean gotLine; // Verification variable
    	String[] returnValue = new String[100]; // Settings max at 100 values
    	String curLine;
    	
    	gotLine = false; // We don't have the settings found yet
    	
    	if( new File(fileName).exists() == false )
    	{
    		return( new String("File not found.ZQX").split("ZQX") );
    	}
		if(splitValue.equals("")) {
			splitValue = "afs4wfa3waawfa3dogrsijkge5ssioeguhwar3awwa3rar";
		}
    	
        try {
        	// Get the line from the file
			FileInputStream fstream = new FileInputStream(fileName);
			BufferedReader in = new BufferedReader(new InputStreamReader(fstream));
			
			while(in.ready()) 
			{
				curLine = in.readLine().toString();
				if(curLine.split("=", 2)[0].equalsIgnoreCase(optionName)) {
					returnValue = new String[100];
					returnValue = curLine.split("=", 2)[1].split(splitValue);
					gotLine = true;
				}
			}
			
			in.close();
			
			// If the line does not exist, create it
			if(!gotLine) 
			{
                returnValue = defaultValue.split(splitValue);
                FileOutputStream out;
                PrintStream p;
                try {
	                out = new FileOutputStream(fileName, true);
	                p = new PrintStream( out );
	                p.println (optionName+"="+defaultValue);
	                p.close();
                } catch (Exception e) {
                	logOutput("Error writing to file");
                }
			}
		}
        catch (Exception e) {
        	logOutput("-=-");
        	logOutput("File input error: "+e.toString());
        	logOutput("File input error: "+e.getStackTrace().toString());
        	logOutput("-=-");
		}
		finally {
		}
		
		return returnValue;
    }

    public Integer getIntSetting( String item, Integer dValue )
    {
    	Integer retVal = Integer.parseInt( getSettingValue(iniFile, item, dValue.toString(), "" )[0]);
    	return retVal;
    }

    public double getDblSetting( String item, double d )
    {
    	double retVal = Double.parseDouble( getSettingValue(iniFile, item, Double.toString(d), "" )[0]);
    	return retVal;
    }

    public Float getFloatSetting( String item, Float dValue )
    {
    	Float retVal = Float.valueOf(getSettingValue(iniFile, item, dValue.toString(), "" )[0]);
    	return retVal;
    }
    
    // generate a new config file
    public static boolean createIniFile()
    {
    	boolean retVal = false;
    	
    	try 
    	{
			FileWriter outFile = new FileWriter(iniFile);
			PrintWriter outP = new PrintWriter(outFile);
			
			outP.println( "flightSpeed=1.1" );
			outP.println( "hoverMultiplier=0.0012410253" );
			outP.println( "hoverBoostMultiplier=0.00555" );
			outP.println( "allowOps=true" );
			outP.println( "flyingEatsFeathers=false" );
			outP.println( "defaultFeatherAmount=500" );
			outP.println( "falldamageWithoutFeather=false" );
			
			outP.close();
			retVal = true;
		} 
    	catch (IOException e) 
    	{
			logOutput( "Error writing to ini file." );
			retVal = false;
			e.printStackTrace();
		}
		
		return retVal;
    }
}
